Add animation_node_notification signal for animation trees

a-johnston opened the pull request:
Closes https://github.com/godotengine/godot-proposals/issues/10253.

Related to https://github.com/godotengine/godot/pull/102398

This adds `animation_node_notification(path: StringName, what: int)` to `AnimationTree` as a generic signal used for node state changes during processing/playback. Notification values are defined for state machine, transition, and one shot nodes. Rather than include several screenshots and an output, I've attached a test project which automatically runs through a few operations and then quits, printing timing information for each step.

[animation_signal_test.zip](https://github.com/user-attachments/files/18916036/animation_signal_test.zip)

a-johnston commented:
Thanks for the links, I hadn't seen those PRs. Do you think it would be worth adding `fading_from_state_changed` to this pr? From the comments, a few things jumped out at me
- Regarding performance concerns, is it more in the sense that adding new signals adds XYZ overhead even if there are no listeners, or have there been related performance issues in the past related to signal use which should be avoided? I can set up a synthetic case that triggers the new emit calls a ton if we want that measured.
- I hadn't considered the option of saving a state machine node for reuse; I'll change the code to not use Resource.name and instead opt for a reverse node -> name map on the tree. Also I don't know if or when this would be useful, but there is an existing odd edge case where the same node resource could be added twice to a tree which would cause `get_node_name` to return the wrong name / cause the wrong node to be updated on a rename callback within the editor. This could be detected; it might be worth adding a warning message or something I'm not sure.

> As I have pointed out in the past with those, it should be implemented so that the ancestor StateMachine returns the nested paths when in Grouped mode [..]. In other words, it must be implemented in such a way that the signal is propagated to the ancestor State as special case for the Grouped mode.

I'll work on adding this. Just to be sure I understand correctly, the changes I'll make would change the example root sm `node_changed` values, assuming InternalStateMachine is set to grouped, to be the following:
```
Start
InternalStateMachine/Anim_Knight_Attack01
InternalStateMachine/Anim_Knight_Attack02
Anim_Knight_Hit01
End
```
Also regarding a comment you left on another PR `the user cannot use the child StateMachine's StateMachinePlayback directly`, it is possible right now to grab the playback for the grouped state machine and listen to the signal from within user code, even if control is not allowed. I'll filter events in the root sm to skip the grouped Start/End node but it'll still be possible for users to directly listen to the grouped sm to detect those states.

TokageItLab commented on scene/animation/animation_blend_tree.cpp:
```diff
@@ -1459,6 +1467,7 @@ void AnimationNodeBlendTree::add_node(const StringName &p_name, Ref<AnimationNod
 	n.position = p_position;
 	n.connections.resize(n.node->get_input_count());
 	nodes[p_name] = n;
+	p_node->set_name(p_name);
```
```suggestion
```

This should never be done. It is the AnimationTree's role to manage the names of AnimationNodes, and AnimationNodes should not contain names.

As explained in some places like https://github.com/godotengine/godot/pull/101564#pullrequestreview-2561174593, an AnimationNode is a resource, so we need to consider the case where multiple AnimationNodes with the same reference are placed in the one blend tree.

a-johnston replied:
I've replaced this with a node to name map which is maintained in parallel to nodes as described in the previous comment

TokageItLab reviewed (changes requested):
> Thanks for the links, I hadn't seen those PRs. Do you think it would be worth adding fading_from_state_changed to this pr?

I think so.

---

As for the `StateMachinePlayback` object, it should be a unique resource for each AnimationTree, so there should be no problem to fire a signal there.

However, if you want to fire a signal in the AnimationNode, signals should not be registered with AnimationNode as explained in the review below.

The simplest implementation would be to fire signal from the AnimationTree with its own BlendTree path. In this case, you can refer to `AnimationTree::set_parameter()` to get the path. At this point, it may be better to separate the method for obtaining its own path from it and make it reusable.

But then, my concern is that each time we have more signals that we want to fire in the AnimationNode, we need to register all of them in the AnimationTree.

So, instead of adding a signal called `oneshot_finished()` to the AnimationTree, we should add a generic abstract signal such as `node_notified(animation_node_path, signal_name, signal_variants)`.

However, that PR should be a separate PR from the `StateMachinePlayback` signal addition.

a-johnston commented:
I've updated the pr to have `node_started`/`node_finished` for state machine nodes rather than `node_changed` so that end of crossfades can be detected via signal. I've also added support for grouped state machines to pass up their signals to the containing playback, including when nested, although I still need to test a few more cases like when state traversal is interrupted by certain calls. In a minute I'll update the op to reflect the current state of the code.

a-johnston commented:
> However, that PR should be a separate PR from the `StateMachinePlayback` signal addition.

This pr is currently sitting at +124 -11 which imo is a good size but if you'd prefer I can break this up into two PRs (maybe a third to rename `fadeing_from_nti` to `fading_from_nti` since that's totally unrelated but was bugging me lol)

TokageItLab commented on doc/classes/AnimationTree.xml:
```diff
@@ -45,6 +45,18 @@
 				Emitted when the [member anim_player] is changed.
 			</description>
 		</signal>
+		<signal name="node_finished">
+			<param index="0" name="node" type="StringName" />
+			<description>
+				Emitted when the [param node] finishes playback. Only emitted for OneShot and StateMachine nodes.
+			</description>
+		</signal>
+		<signal name="node_started">
+			<param index="0" name="node" type="StringName" />
+			<description>
+				Emitted when the [param node] starts playback. Only emitted for OneShot and StateMachine nodes.
+			</description>
+		</signal>
```
I suggest that it should be a generic signal as explained above.

> So, instead of adding a signal called oneshot_finished() to the AnimationTree, we should add a generic abstract signal such as node_notified(animation_node_path, signal_name, signal_variants).

For example, what happens with NodeTransition? Don't we want a signal when BlendSpace is in DC mode and the values switch discretely? What should we use when firing a signal from a custom AnimationNode? I suppose it is not good way to add a new signal to the AnimationTree for them every time.

TokageItLab replied:
However, the concern remains only that generic signals use strings as signal identifiers. Another idea is to pass some kind of Callable object from the AnimationNode to the AnimationTree.

a-johnston replied:
What sort of callable do you imagine, and would it make more sense to use an enum rather than bare strings? Also are there other examples of notification style multipurpose callbacks?

I'll think about it more but from my perspective a lot of these node notify events would be directly caused by user code, such as requesting a transition or changing the blendspace mode, and could be handled by an animation controller script. In comparison, I think `node_started` and `node_finished` would be harder to implement without extra coupling or polling of the tree's state. That said, it probably won't get to the point where the generic signal gets too expensive from unnecessary connections, so the question of which things should be node notified emits or not doesn't really matter. It's also the case that this doesn't need to be decided without some future use case in mind - the state machine signal addresses the proposal use case (and my use case) on its own.

Either way I'll give the generic signal a shot and also break off the state machine changes.

a-johnston replied:
Ok, this has been updated to use a generic signal `node_notify` and the other class was split into its own pr

a-johnston committed:
Add node_notify signal for animation trees

TokageItLab commented on doc/classes/AnimationNode.xml:
```diff
@@ -244,5 +244,11 @@
 		<constant name="FILTER_BLEND" value="3" enum="FilterAction">
 			Paths matching the filter will be blended (by the blend value).
 		</constant>
+		<constant name="NOTIFY_STARTED" value="0" enum="NotifyReason">
+			The node has started playback.
+		</constant>
+		<constant name="NOTIFY_FINISHED" value="1" enum="NotifyReason">
+			The node has finished playback.
+		</constant>
```
Using ENUM is the right direction to go. However, for consistency with something similar to this that already exists in Godot, we recommend the signal name NOTIFICATION_XXX instead of NOTIFY_XXX.

a-johnston replied:
Hm, would `NOTIFICATION_XYZ` cause confusion with the existing notification stuff? I preemptively changed the enum name as well but that causes the c# build to complain about hiding `Notification` so that at least can't be shared. Maybe `EVENT_STARTED` etc with enum name `Event`?

TokageItLab replied:
How about `ANIMATION_NODE_NOTIFICATION_XXX` ?

TokageItLab commented on scene/animation/animation_blend_tree.h:
```diff
@@ -411,6 +411,7 @@ class AnimationNodeBlendTree : public AnimationRootNode {
 	};
 
 	RBMap<StringName, Node, StringName::AlphCompare> nodes;
+	RBMap<AnimationNode *, StringName> node_to_name;
```
I believe this approach is redundant. At the very least, storing the same String in two places is not the right direction, and we must do indirect mapping, such as storing only the index, and the entity as a String must be only one.

a-johnston replied:
isn't StringName only storing the unified data pointer within either map, so there is just one backing string? this is just to avoid the linear cost of only having `nodes` to determine a name in the current context. ideally there would just be a bimap template available, but alternatively the name could also be copied to `NodeState` at the same time that the base path and parent are set.

TokageItLab replied:
I am saying that the concatnated path to the AnimationNode already exists in the property_reference_map, so you don't need to create a new map and just retrieve it.

a-johnston replied:
Do you think it would be better to parse out the final part of the path or emit the full concatenated path?

TokageItLab replied:
The signal must return the full path to the AnimationNode. As I have said many times, since the AnimationNode is a Resource, the same reference AnimationNodes can be placed in the same hierarchy. To distinguish between them, the user must know the full path.

a-johnston replied:
Ah yes that makes sense. I was only considering blend trees as roots but of course the names alone would break if they were nested.

TokageItLab commented on scene/animation/animation_blend_tree.cpp:
```diff
@@ -696,6 +696,10 @@ AnimationNode::NodeTimeInfo AnimationNodeOneShot::_process(const AnimationMixer:
 	set_parameter(fade_in_remaining, cur_fade_in_remaining);
 	set_parameter(fade_out_remaining, cur_fade_out_remaining);
 
+	if (cur_active != (bool)get_parameter(active)) {
+		_notify_tree(cur_active ? NOTIFY_FINISHED : NOTIFY_STARTED);
+	}
+
```
Since the states of internal_active and current_active are not synchronized, it might be better to have an explanation in the documentation that they are “visual” and different from the "internal state".

To give a detailed example, in the case of trying to manage Node state with user scripts, OneShot has been internally terminated already when fading out, so restarting AnimationTree should cause a synchronization discrepancy if you are relying on signals.

It may be possible to add a signal STATE_CHANGED as well as documentation to solve this.

a-johnston replied:
I've added `NOTIFICATION_FADEOUT_STARTED` to allow this to be tracked by a user. Currently only one shot nodes emit it since state machines only have crossfade on their internal nodes and that can be fully tracked by the node started/finished events. It also only emits specifically when there is non-zero fadeout, so a bit different from if it was strictly a version of this line but for internal_active.

TokageItLab commented on scene/animation/animation_tree.h:
```diff
@@ -54,6 +54,12 @@ class AnimationNode : public Resource {
 		FILTER_BLEND
 	};
 
+	enum AnimationNodeNotification {
+		ANIMATION_NODE_NOTIFICATION_STARTED,
+		ANIMATION_NODE_NOTIFICATION_FINISHED,
+		ANIMATION_NODE_NOTIFICATION_FADEOUT_STARTED
+	};
+
```
I think the overall implementation is headed in the right direction.

My concern is where to define the enum for signal, maybe it should be defined separately for each AnimationNode and returned as an int in the AnimationTree like:

animation_blend_tree.h
```
	enum AnimationNodeNotificationOneShot {
		ANIMATION_NODE_NOTIFICATION_ONESHOT_STARTED,
		ANIMATION_NODE_NOTIFICATION_ONESHOT_FINISHED,
		ANIMATION_NODE_NOTIFICATION_FADEOUT_FINISHED_INTERNAL, // Fadeout started.
	};
~~~~~~
	enum AnimationNodeTransitionNotification {
		ANIMATION_NODE_NOTIFICATION_TRANSITION_STARTED,
		ANIMATION_NODE_NOTIFICATION_TRANSITION_FINISHED,
	};
```

a-johnston replied:
Yeah I was thinking about this. The meaning of some of the notifications are frustratingly similar but I agree it's probably best to separate them for clarity and ease of documentation.

a-johnston replied:
Revisiting this, I think actually each of the oneshot, state machine, and transition nodes could use `STARTED`, `FINISHED`, and `FINISHED_INTERNAL` with near identical semantics. At least for next version of this pr I'm planning on having those three constants defined on AnimationNode but use int for the signal/helper method so that AnimationNode subclasses can still define their own notification values and document them individually.

I do have a question as this version of the pr has `FADEOUT_STARTED` as opposed to `FINISHED_INTERNAL` and I've kept the code to only emit that value when there is nonzero crossfade; would you want `FINISHED_INTERNAL` to fire in cases where there is no fadeout?

TokageItLab replied:
I am beginning to think that FINISHED_INTERNAL is a little too difficult for users who do not touch the core, so FADEOUT_STARTED might be fine.

> I do have a question as this version of the pr has FADEOUT_STARTED as opposed to FINISHED_INTERNAL and I've kept the code to only emit that value when there is nonzero crossfade; would you want FINISHED_INTERNAL to fire in cases where there is no fadeout?

If there is no crossfade, it is better not to fire. However, if the implementation is too difficult, it is fine to fire it, but perhaps you should note that it fires even if there is no fade-out in the documentation.

a-johnston replied:
Ok, I had to have this on the backburner for a bit but I've updated the pr with per-class enums. I ended up using unnamed enums with manually-made-unique values to be more similar to notification but that can easily be changed to whatever.

I think my only issue with the current implementation is how verbose qualifying the enum is but maybe either the name could be changed or it could be bound as a new globalscope constant. I also noticed a while ago that a state machine with no valid path to the end node never actually finishes (existing behavior; a one shot triggering such a sm will not finish) and included that in the documentation although I'm not sure if it's really the desired behavior?

TokageItLab commented on scene/animation/animation_tree.cpp:
```diff
@@ -991,6 +1001,7 @@ void AnimationTree::_bind_methods() {
 	ADD_PROPERTY(PropertyInfo(Variant::NODE_PATH, "anim_player", PROPERTY_HINT_NODE_PATH_VALID_TYPES, "AnimationPlayer"), "set_animation_player", "get_animation_player");
 
 	ADD_SIGNAL(MethodInfo(SNAME("animation_player_changed")));
+	ADD_SIGNAL(MethodInfo(SceneStringName(node_notification), PropertyInfo(Variant::STRING_NAME, "node"), PropertyInfo(Variant::INT, "what")));
```
```suggestion
	ADD_SIGNAL(MethodInfo(SceneStringName(animation_node_notification), PropertyInfo(Variant::STRING_NAME, "animation_node_path"), PropertyInfo(Variant::INT, "what")));
```
"node" should be avoided because it is confusing with Node/Node2D/Node3D classes.

TokageItLab commented on scene/animation/animation_tree.h:
```diff
@@ -181,6 +181,8 @@ class AnimationNode : public Resource {
 
 	void _validate_property(PropertyInfo &p_property) const;
 
+	void _notify_tree(int p_what);
```
```suggestion
	void _animation_tree_notify(int p_what);
```
The “tree” should also be avoided as far as possible to prevent confusion between SceneTree and AnimationTree.

TokageItLab commented on scene/animation/animation_blend_tree.h:
```diff
@@ -157,6 +164,8 @@ class AnimationNodeOneShot : public AnimationNodeSync {
 	StringName fade_out_remaining = "fade_out_remaining";
 	StringName time_to_restart = "time_to_restart";
 
+	void _maybe_notify(bool p_active, bool p_internal_active, double p_fade_in_remaining);
```
```suggestion
	void _check_notify(bool p_active, bool p_internal_active, double p_fade_in_remaining);
```
Indicate what exactly the function is doing.

TokageItLab reviewed (changes requested):
The most implementation looks good, but the function names need further polishing.

TokageItLab commented on scene/animation/animation_node_state_machine.h:
```diff
@@ -116,6 +116,11 @@ class AnimationNodeStateMachine : public AnimationRootNode {
 		STATE_MACHINE_TYPE_GROUPED,
 	};
 
+	enum {
+		ANIMATION_NODE_NOTIFICATION_STATE_MACHINE_STARTED = 6,
+		ANIMATION_NODE_NOTIFICATION_STATE_MACHINE_FINISHED = 7
+	};
+
```
```suggestion
	enum {
		ANIMATION_NODE_NOTIFICATION_STATE_MACHINE_STARTED = 6,
		ANIMATION_NODE_NOTIFICATION_STATE_MACHINE_FINISHED = 7
	};
```
Managing IDs with sequential numbers across classes is not recommended.

If you really want to do this, at least for example, have the ID for each AnimationNode in an enum on the AnimationTree side, and multiply it by 1000 to add it, so that the AnimationNode can have signals from 0 to 999.

However, I think it is still complicated, so I think it is easiest for the users to understand by passing class names to the signals.

a-johnston replied:
I mean at that point would it be best to just have the full enum defined on the base `AnimationNode`? The per signal documentation wouldn't be on the specific node that emits it but at least it would all be in one place and the values would be unique. I think it would be very likely for the user to make the mistake of comparing against a normal string, and it feels weird passing the parameter as a type when it's already a part of how the enum is defined (currently twice).

TokageItLab replied:
```gdscript
if node_type == &"AnimationNodeStateMachine" and what == AnimationNodeStateMachine.ANIMATION_NODE_NOTIFICATION_STATE_MACHINE_FINISHED:
```
Yes, this is recommended with
```
ADD_SIGNAL(MethodInfo(SceneStringName(animation_node_notification), PropertyInfo(Variant::STRING_NAME, "animation_node_path"), PropertyInfo(Variant::STRING_NAME, "animation_node_class_name"), PropertyInfo(Variant::INT, "what")));
```

a-johnston replied:
Were there issues/regrets with the notification guidance of `You can make your own, but don't use the same numbers as other notifications in other nodes`? I don't think having two parameters here makes it per se safer when the enum values are allowed to overlap compared to if the enum values were not allowed to overlap to begin with. While there is some overhead to checking existing values (ie search the codebase for `ANIMATION_NODE_NOTIFICATION_[A-Z_]+ = \d+`) I don't think it's too out of question for how infrequently it would happen in this case.

I can definitely just give each class a range of 100 or 1000 values to make it easier though. If that "assigned" range ever ran out adding new values without changing old ones would still possible since values don't need to be contiguous, but probably it would be worth figuring out something different entirely for handling whatever would cause that.

TokageItLab replied:
First, I think we can agree on the following two points.

- Signals should be defined for each AnimationNode class
- The number should not be a sequential number across classes

Therefore, overlapping numbers per class, e.g., `ANIMATION_NODE_NOTIFICATION_ONESHOT_STARTED = 0` and `ANIMATION_NODE_NOTIFICATION_STATE_MACHINE_STARTED = 0`,  is not a problem.

However, if `animation_node_notification()` only returns int and pathname, there is a problem with users cannot distinguish between them. So in order to solve that problem, animation_node_notification must return the class_name.

a-johnston replied:
I agree with the first point but not the second. I think having repeat notification values and relying on an extra path to be passed is error prone compared to making the enum values distinct across classes, although at least with the current 8 values "started" and "finished" would always be 0 and 1 so if someone tab completed the wrong one of those it would still work (as long as it was also qualified by the type or path). That's why I was asking if more generally there was a consensus about how notifications are split across classes being bad, since that's the closest inspiration for how the signal is defined and used at the moment. It took me only a few seconds to find all the unique `NOTIFICATION_` definitions so I don't think it's too unreasonable for animation notifications to have distinct values.

TokageItLab replied:
The second problem, as raised earlier, is that the sequential numbering is broken when signals are added or deleted. This is unacceptable.

The long enum names are unavoidable in order to conform to Godot's naming conventions, and this is not the only area where long enums are used.

a-johnston replied:
> The second problem, as raised earlier, is that the sequential numbering is broken when signals are added or deleted. This is unacceptable.

By sequential numbering do you mean specifically having the enum values 0, 1, 2 etc for each class without gaps? In my experience when constants like these are deprecated/removed, a hole is left in the numbering so that serialized data still maps to the correct name. For adding enums, I do think it's best to allocate a large range per class (so its enum values wouldn't have gaps, but also wouldn't start at 0) if mostly for convenience, but even without that I don't think there's any technical problem with having gaps in a given class's notification values (`Node` does this, for example).

> I don't know what you mean. The type-safe check is only done once by comparing class name StringName

I should have been clearer. You're right the `node_type` is the part actually doing the type check in that case but I was talking about literally specifying the type. For example in the following we want to check that a state machine has finished:
```gdscript
if node_type == &"AnimationNodeStateMachine" and what == AnimationNodeStateMachine.ANIMATION_NODE_NOTIFICATION_STATE_MACHINE_FINISHED:
```
The type gets specified in the first equality, the classname itself of course, and then the enum name. Having the name as a separate parameter would be useful if we wanted to define the enum on a common base class or wanted each class to be zero-indexed but I don't think either is correct/necessary here. IMO, it doesn't need to be in the enum name because it always is qualified by the class preceding it, but I don't feel as strongly about this.

I'd prefer to cut out the name there though than to remove `ANIMATION_NODE_` as it might make people think it gets used with `_notification` (also sort of the reason I originally went with a different word entirely, although that too does not scale well in a large codebase). Maybe `AnimationNodeStateMachine.ANIMATION_EVENT_STARTED` would be better if it was acceptable to use a different word than notification.

TokageItLab commented on scene/animation/animation_tree.cpp:
```diff
@@ -991,6 +997,7 @@ void AnimationTree::_bind_methods() {
 	ADD_PROPERTY(PropertyInfo(Variant::NODE_PATH, "anim_player", PROPERTY_HINT_NODE_PATH_VALID_TYPES, "AnimationPlayer"), "set_animation_player", "get_animation_player");
 
 	ADD_SIGNAL(MethodInfo(SNAME("animation_player_changed")));
+	ADD_SIGNAL(MethodInfo(SceneStringName(animation_node_notification), PropertyInfo(Variant::STRING_NAME, "animation_node_path"), PropertyInfo(Variant::INT, "what")));
```
```suggestion
	ADD_SIGNAL(MethodInfo(SceneStringName(animation_node_notification), PropertyInfo(Variant::STRING_NAME, "animation_node_path"), PropertyInfo(Variant::STRING_NAME, "animation_node_class_name"), PropertyInfo(Variant::INT, "what"), PropertyInfo(Variant::PACKED_INT64_ARRAY, "sub_indices")));
```
Also I recommend adding `sub_indices` so that the transition source and destination can be passed.

If connect_deferred is used, they may not match when looking at the current Transition state by retrieving the AnimationTree parameters from within the signal. So it is safe to pass them as signal arguments.

Ideally, vararg should be used, but since signals cannot use vararg, so this is a compromise solution.

a-johnston replied:
I was thinking about this too, although initially I was leaning towards having an ultra generic `Variant extra` parameter since no obvious subset of state existed between all animation node types. `int[]` probably makes 
sense as the most specific type but I am very uncertain about effectively making adhoc marshaling/encoding a part of the signature and for the semantics to be completely disjoint between node types. It's also a bit of a slippery slope deciding what data gets to be included considering the premise is that the user needs to access a snapshot in time of the state. But then a more general data container would invite a situation similar to [this](https://sampruden.github.io/posts/godot-is-not-the-new-unity/) especially in light of this being a generic signal which is likely being immediately filtered on `what` in the callback.

In c# this question (as well as the enum problem) would be addressed by `EventArgs` but I think the only way to achieve something similar with the godot api would be by creating and passing a `RefCounted` subclass with each emit. This seems to me too expensive to opt all animation tree users into.

Currently I think the best thing would be to have these notifications in places where state is meaningful (ie `TRANSITION_STARTED` having the current and requested states both be populated) and rely on users treating the callback as the correct time to check/process the animation tree state. If there's such a case where the processing absolutely needs to be deferred, an intermediate callback could read the state and pass that to the deferred call.

a-johnston commented:
Forgot to push up some changes the other day. The most recent changes fix the multi-emit for nested state machines and the error which could happen when starting a state machine. I also noticed `playing` on state machine is only reset to `false` when the current state does not exist and when the state machine is being reset. Should it be set to false when the state machine finishes normal playback? I could make a helper which also emits that finished notification.

TokageItLab commented:
In my opinion, I cannot fully agree unless you add "animation_node_class_name" and "sub_indices", as the above enum indexing issue is still there.

Any opinions from the animation team on this? @SaracenOne @lyuma @fire

a-johnston commented:
That's fair. Were that part internal I would just make the change, but since it's part of the public api I think getting other opinions on how it should be set up is good.

a-johnston commented:
I'll try changing this to use InputEvent as inspiration rather than notification. I think passing polymorphic refcounted instances best matches the use case and extensions for this pr, but wrote it off before due to the potential cost combined with how frequently the AnimationTree node is already used in existing projects. I'll actually benchmark it though and if it is too expensive it can be mitigated by making the feature opt-in (which might be nice to do regardless of performance).

a-johnston committed:
change to use animation node events

TokageItLab commented:
Resource seems quite overdesigned, since this kind of hook is not reused as often as shader. It should at least be a Callable.

I feel it would be cleaner if we could pass a Callable, but in any case we need to pass the ClassName as a Callable argument, so it would not be much different from having a signal with several arguments, as I suggested above.

a-johnston commented:
> For example: [..]

Would the intended usage of such an approach (in user code) be similar to:

```gdscript
if notification.get_method() == &"animation_node_transition_started":
    var args = notification.get_bound_arguments()
    print("From %d to %d" % args)
```

as a sort of makeshift untyped (from the user POV) tuple with a name? I would definitely prefer passing the name and an array/dict of values compared to this, which is what you were getting at. In either case I dislike forcing the user to match against a magic string and to use a magic int (in the case of multiple binds) to get the value they want. Users could also do something like `args[-1]` and as a result any values that get added over time would be breaking. I think it is also significant that names and (static) types are lost here compared to either 1) the trivial case of passing minimal data and forcing the user to immediately query what they need or 2) having fully structured events.

> BTW, the AnimationNodeEvent resource approach has AnimationNode as a member, which as I have explained several times, is quite dangerous from a standpoint of the pointer safety (one AnimationNode entity can be referenced by multiple AnimationTree instances, so we should not allow easy access to the AnimationNode in the process), therefore it should never be approved.

The previous issue you had explained was related to what kind of state can exist on the animation node instances due to them being resources (which was definitely my mistake not realizing what was allowed within the animation system), rather than users being disallowed "easy access" to these nodes in the first place. This seems like a strange position to take when for example the most obvious (only?) public api way to convert from `int p_index` in your example to a usable name uses the AnimationNode instance.

> In any case, rather than implementing a hook with a Callable or creating a Resource, it should be simpler and easier for the user to understand to just pass a class name string. The Godot's [philosophy](https://docs.godotengine.org/en/latest/contributing/development/best_practices_for_engine_contributors.html) is that the solution to a purpose should be minimal.

Frankly, it is frustrating to read this as from my point of view the earlier approach for this pr was dead simple and very easy to extend as use cases arose but faced objections which required added complexity/redundancy for questionable benefit. When I disagreed with or tried to clarify the objections it felt you were only reading every other line of my comment or something. Then later I changed the pr to allow for the use cases you wanted, and then came the pushback with complexity and use of Resource, even though considerations were made for performance and it's a base class for a notable transient type - InputEvent.

That being said, I respect that I'm not entitled to have my pr merged into your code. I don't really want my name attached to the interfaces you've suggested changing this to and I'm not hugely interested in wasting more time on this so I'll be closing this pr. <https://github.com/godotengine/godot/pull/102398> on its own addresses the proposal's use case, although I personally have use for these signals so I'll rebase this to my own dev branch. Others are free to use this code as a starting point for their own prs/use. Sorry for the essay this comment turned into.

TokageItLab commented:
You may have misunderstood some sentence; in my opinion, the signal approach was fine before you changed to the Event resource approach, except for the arguments. I suggested passing a StringName of class_name which is guaranteed to be compatible compared to an enum which could be broken over time/version, but you disagreed.

Passing the Transition's port number through the sub_indice array does not imply access to the AnimationNode; if you think of the Transition as a StateMachine, it simply indicates which state it has transitioned to.

So I think it is salvageable by revert to approach that uses signal and pass class_name as StringName argument.

a-johnston commented:
> You may have misunderstood some sentence; in my opinion, the signal approach was fine before you changed to the Event resource approach, except for the arguments. I suggested passing a StringName of class_name which is guaranteed to be compatible compared to an enum which could be broken over time/version (depends on the addition of AnimationNode and signals), but you disagreed.

I haven't misunderstood your position, I have only disagreed. I've already said as much before but I consider use of a magic string to be more brittle than other options and I disagreed with your characterization of when enums work or do not work. The issues you described with enums are a direct result of what you were requiring the values to be and didn't reflect how enums are used by _notification or more broadly. The event classes are in my opinion the best design for the requirements you at various points described:

- having a single generic signal for all animation nodes
- allowing extra data to be included for easy use with deferred callbacks
- distinct values which will not break/shift as things are changed in the future

> Passing the Transition's port number through the sub_indice array does not imply access to the AnimationNode; if you think of the Transition as a StateMachine, it simply indicates which state it has transitioned to. Since Transition's current_index is an AnimationTree parameter which is guaranteed to be one per AnimationTree instance (AnimationTreeInstance n vs n AnimationTreeParameter), it is clearly different from accessing the internal data of an AnimationNode resource (AnimationTreeInstance n vs 1 AnimationNodeResource).

My point was that users would need to either 1) maintain a duplicate map of index to state name or 2) call `AnimationNode::get_input_name` to get a value to either compare against or pass as a transition request. The code could be modified to also include `prev_state` in the parameters, or to accept ints for transition requests, but that's not currently the state of things. In any case, given the path it's trivial to get access to the resource anyways. If access and possible mutation during process is really a problem then it is currently already a problem.

My primary issue with `sub_indices` is that it is shoehorned into the signature and would be used by exactly one class and have zero other use within this pr. You mentioned other hypothetical use cases, but they all involve overloading the meaning of the `int[]` which I consider a significant issue with the proposed interface. A more general data structure could be used but, compared to structured events, more magic values are introduced into user code and static typing is impossible.

TokageItLab commented:
I disagree with your opinion that class_name is a magic string, since class_name is frequently used elsewhere. You seem to be hung up on the enum traversing the AnimationNode, but I have decided that an enum + class_name independent of each AnimationNode is preferable, since it is obviously more maintainable.

> My point was that users would need to either 1) maintain a duplicate map of index to state name or 2) call AnimationNode::get_input_name to get a value to either compare against or pass as a transition request. The code could be modified to also include prev_state in the parameters, or to accept ints for transition requests, but that's not currently the state of things. In any case, given the path it's trivial to get access to the resource anyways. If access and possible mutation during process is really a problem then it is currently already a problem.

This may be a reasonable solution by passing the StringName of the port name as sub_info instead of sub_indices, but it would also be fine to make the index available to the Transition request.

In any case, passing an Event resource or dictionary is not allowed due to performance concerns. In the past, for the same reason, `AnimationNodeExtention` was implemented using the approach of packing `struct NodeTimeInfo` into a `PackedIntArray` by https://github.com/godotengine/godot/pull/99181.

If struct is implemented in GDScript, I believe the best approach would be to use struct. However, even in that case, the class_name should still be passed as StringName, not passing a reference to the AnimationNode resource directly.

a-johnston commented:
You're right, they're not true magic values. Specifically I meant the use of literals with little to no static checking possible, which is my main issue with the recommendations. I don't mind class name that much (besides the line length), although it's surprising the resource is so off limits. I don't like passing array/dict for additional data though and forcing users to know which indices/keys mean what for which animation nodes. I'm not sure what you mean by the enum traversing the animation nodes; I think we only disagreed on if each enum should be zero indexed or offset such that it is more easily associated with its class (which overlaps in function with passing class name). I hope it comes across that for each of the approaches considered in this thread I have implemented the changes and written short scripts from the user perspective to test the ergonomics.

The reason I'm surprised about the node instances is that they offer much more than just the class name and they're already controlled/easily accessible by the user. If corrupting/changing state during process is an issue, that should be itself addressed. For example, warning/erroring when calling a public non-const method while the state pointer is set. Similarly, that pr looks like it's fixing issues inherent to process state being a part of the resource state rather than for example having the process calls be entirely functional or having a stack to prevent clobbering (although it's a huge pr so I skimmed; sorry if I missed something).

Benchmarking a few hundred instances, they're both very fast but using a resource was on average 1.6 microseconds slower than using packed array. Calling `emit_signal` with nothing connected is more expensive than creating + setting up either, and adding a simple callback makes it an order of magnitude slower than either option. And of course it can be opt-in on a node by node basis regardless of implementation anyways. The packed array is still a dynamically sized cow vector and isn't free; to me the performance difference is more than appropriate for the advantages of resource. Structs could be a good alternative, especially if they offered a more concise way of declaring the type and its fields than this.
